### DISCUSSION TITLE
AbstractConcreteBase - dir() of concrete subclass shows sibling columns

### DISCUSSION THREAD

**Discussion by John-Sias** [sqlalchemy/sqlalchemy#8402] [Usage Questions] [answered]
I'm working on 1.4.40 (but this behavior is the same on 1.3.24), documenting an ORM we use in a somewhat "live-interactive" mode, which uses AbstractConcreteBase for class inheritance hierarchy. I've found that the Python dir() of concrete subclasses of the abstract base shows (rather misleadingly) columns from sibling concrete subclasses after configure_mappers() is invoked.

(For example, in the simple example from the docs, Engineer has a manager_data attribute and Manager has an engineer_info attribute; both have both, just as parent Employee does)

This is at least counter-intuitive to me -- and perhaps a bug? Or am I missing something here?

Thanks for any thoughts!

John

```
from sqlalchemy import Column, Integer, String, create_engine
from sqlalchemy.orm import configure_mappers, declarative_base
from sqlalchemy.ext.declarative import AbstractConcreteBase

Base = declarative_base()

class Employee(AbstractConcreteBase, Base):
    pass

class Manager(Employee):
    __tablename__ = "manager"
    id = Column(Integer, primary_key=True)
    name = Column(String(50))
    manager_data = Column(String(40))

    __mapper_args__ = {
        "polymorphic_identity": "manager",
        "concrete": True,
    }


class Engineer(Employee):
    __tablename__ = "engineer"
    id = Column(Integer, primary_key=True)
    name = Column(String(50))
    engineer_info = Column(String(40))

    __mapper_args__ = {
        "polymorphic_identity": "engineer",
        "concrete": True,
    }

def dump_dir_and_cols(c):
    print(c,'\n'+40*'-'+'\ndir(): ',
          [p for p in dir(c) if not p.startswith('_')],'\n__mapper__.c: ',
          'x' or list(c.__mapper__.c))

print('Before configure_mappers:')

for c in [Employee, Manager, Engineer]:
    dump_dir_and_cols(c)

configure_mappers()

print('After configure_mappers:')

for c in [Employee, Manager, Engineer]:
    dump_dir_and_cols(c)

```

Output:

```
Before configure_mappers:
Employee 
----------------------------------------
dir():  ['metadata', 'registry'] 

Manager 
----------------------------------------
dir():  ['id', 'manager_data', 'metadata', 'name', 'registry'] 

Engineer 
----------------------------------------
dir():  ['engineer_info', 'id', 'metadata', 'name', 'registry'] 

After configure_mappers:
Employee 
----------------------------------------
dir():  ['engineer_info', 'id', 'manager_data', 'metadata', 'name', 'registry', 'type'] 

Manager 
----------------------------------------
dir():  ['engineer_info', 'id', 'manager_data', 'metadata', 'name', 'registry', 'type'] 

Engineer 
----------------------------------------
dir():  ['engineer_info', 'id', 'manager_data', 'metadata', 'name', 'registry', 'type'] 
```

**Comment by zzzeek** [comment, marked as answer]
> I'm working on 1.4.40 (but this behavior is the same on 1.3.24), documenting an ORM we use in a somewhat "live-interactive" mode, which uses AbstractConcreteBase for class inheritance hierarchy. I've found that the Python dir() of concrete subclasses of the abstract base shows (rather misleadingly) columns from sibling concrete subclasses after configure_mappers() is invoked.
> 
> (For example, in the simple example from the docs, Engineer has a manager_data attribute and Manager has an engineer_info attribute; both have both, just as parent Employee does)
> 
> This is at least counter-intuitive to me -- and perhaps a bug? Or am I missing something here?


well it's kind of both, AbstractConcreteBase is not very good and it maps to a big UNION of all your subclasses.  so it naturally gains attributes for all those classes, which then all derive from it and therefore have all those attributes, and it's a mess.     it's not like there isn't likely some better way for the use case to function, but abstractconcretebase is kind of a workaround for something that doesn't have native support within mappers right now.

typical solution: dont use AbstractConcreteBase.  if you have concrete classes, query them individually.  if you want to be able to query "the hierarchy", that is, all the subclasses at once, use a different form of inheritance (single or joined), which both can emit more efficient forms of queries.


> 
> Thanks for any thoughts!
> 
> John
> 
> ```
> from sqlalchemy import Column, Integer, String, create_engine
> from sqlalchemy.orm import configure_mappers, declarative_base
> from sqlalchemy.ext.declarative import AbstractConcreteBase
> 
> Base = declarative_base()
> 
> class Employee(AbstractConcreteBase, Base):
>     pass
> 
> class Manager(Employee):
>     __tablename__ = "manager"
>     id = Column(Integer, primary_key=True)
>     name = Column(String(50))
>     manager_data = Column(String(40))
> 
>     __mapper_args__ = {
>         "polymorphic_identity": "manager",
>         "concrete": True,
>     }
> 
> 
> class Engineer(Employee):
>     __tablename__ = "engineer"
>     id = Column(Integer, primary_key=True)
>     name = Column(String(50))
>     engineer_info = Column(String(40))
> 
>     __mapper_args__ = {
>         "polymorphic_identity": "engineer",
>         "concrete": True,
>     }
> 
> def dump_dir_and_cols(c):
>     print(c,'\n'+40*'-'+'\ndir(): ',
>           [p for p in dir(c) if not p.startswith('_')],'\n__mapper__.c: ',
>           'x' or list(c.__mapper__.c))
> 
> print('Before configure_mappers:')
> 
> for c in [Employee, Manager, Engineer]:
>     dump_dir_and_cols(c)
> 
> configure_mappers()
> 
> print('After configure_mappers:')
> 
> for c in [Employee, Manager, Engineer]:
>     dump_dir_and_cols(c)
> ```
> 
> Output:
> 
> ```
> Before configure_mappers:
> Employee 
> ----------------------------------------
> dir():  ['metadata', 'registry'] 
> 
> Manager 
> ----------------------------------------
> dir():  ['id', 'manager_data', 'metadata', 'name', 'registry'] 
> 
> Engineer 
> ----------------------------------------
> dir():  ['engineer_info', 'id', 'metadata', 'name', 'registry'] 
> 
> After configure_mappers:
> Employee 
> ----------------------------------------
> dir():  ['engineer_info', 'id', 'manager_data', 'metadata', 'name', 'registry', 'type'] 
> 
> Manager 
> ----------------------------------------
> dir():  ['engineer_info', 'id', 'manager_data', 'metadata', 'name', 'registry', 'type'] 
> 
> Engineer 
> ----------------------------------------
> dir():  ['engineer_info', 'id', 'manager_data', 'metadata', 'name', 'registry', 'type'] 
> ```

**Reply by zzzeek** [reply]
this patch shows how to make the attributes be what you expect them to be.  however, queries don't work anymore.     the main issue is that the ORM has no concept of a mapper that has no table of its own, even though it has the ability to query for a larger set of tables than its own.   for abstractconcretebase to be any good, a whole new feature would need to be added into the Mapper construct and throughout the ORM.    The problem is that concrete inheritance is just not a very popular pattern and the effort has never been warranted.

```diff
diff --git a/lib/sqlalchemy/ext/declarative/extensions.py b/lib/sqlalchemy/ext/declarative/extensions.py
index 87fc702cfb..f8d48b0632 100644
--- a/lib/sqlalchemy/ext/declarative/extensions.py
+++ b/lib/sqlalchemy/ext/declarative/extensions.py
@@ -281,10 +281,11 @@ class AbstractConcreteBase(ConcreteBase):
         def mapper_args():
             args = m_args()
             args["polymorphic_on"] = pjoin.c[discriminator_name]
+            args["include_properties"] = list(pjoin.primary_key) + [discriminator_name]
+            args["with_polymorphic"] = ('*', pjoin)
             return args
 
         to_map.mapper_args_fn = mapper_args
-
         m = to_map.map()
 
         for scls in cls.__subclasses__():

```

**Reply by John-Sias** [reply]
Wow, thanks! I'll try that out.

I have a large existing database with type/id key relationships connecting "thing" (person, org_type_1, org_type_2 ... org_type_n ...) tables to attribute tables (phone number, status, employment relationships, organizational relationships). I have played a bit with this style of polymorphism to work on queries like "what types of orgs employ what types of people" and, as you've noted, had some difficulty with the model. Nonetheless, the basic framework has worked fairly well for me to encapsulate the model.

Thanks -- appreciate the clear and fast answers and all the work that's gone into this fantastic library!

John

**Reply by zzzeek** [reply]
im seeing if I can make the patch above do most of the things we want.    definitely not for 1.4.   for 2.0 it works a bit differently so maybe i will call it abstractconcretebasev2 or something

**Reply by zzzeek** [reply]
i set up #8403 for that

**Reply by John-Sias** [reply]
Well, I could simply solve my problem by overriding __dir__ or by building something else to use for interactive model exploration. It may be just as well to leave whatever change to a more comprehensive approach to concrete inheritance -- if interest warrants. Nonetheless, thanks much!
